### PR TITLE
Utils for modyfing content. Grouping blocks. Grouping of extension methods. Companion object rendering

### LINKED ISSUE #60: Companion object should be first class members
Goes for both class members and package members.

### PR DESCRIPTION
**Pull request by pikinier20**
image

closes #60 

image

image

### CODE REVIEW

**Comment by pikinier20** [comment]
closes #60

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala [hunk 1]
```diff
@@ -99,14 +104,205 @@ class ScalaPageCreator(
...
+                        )
+                    )
+                case _ => head
+            }
+            case _ => modifiedContentNode
+        }
+
+    private def getContentGroupWithParents(root: ContentGroup, condition: ContentGroup => Boolean): Seq[ContentGroup] = {
+        def getFirstMatch(list: List[ContentNode]): Seq[ContentGroup] = list match {
+            case (head: ContentNode) :: (tail: List[ContentNode]) => head match {
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/ScalaPageCreator.scala]: Are these types necessary? Type ascriptions in pat-mat are type tests, not simple annotations, so I'd avoid adding them unless they are necessary.

File: src/main/scala/dotty/dokka/tasty/SyntheticSupport.scala [hunk 2]
```diff
@@ -33,4 +33,10 @@ trait SyntheticsSupport:
     given dotc.core.Contexts.Context = r.rootContext.asInstanceOf
     val sym = rsym.asInstanceOf[dotc.core.Symbols.Symbol]
     sym.is(dotc.core.Flags.Extension)
-  }
\ No newline at end of file
+  }
+
+  def getExtendedSymbol(d: Symbol): Option[ValDef] = 
+    Option.when(hackIsExtension(self.reflect)(d))(
+      if(d.name.endsWith(":")) d.tree.asInstanceOf[DefDef].paramss(1)(0)
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/tasty/SyntheticSupport.scala]: there should be a `.isLeftAssoc` method exposed in Tasty-Reflect but we forgot to do it, can we change this line to something line `hackIsLeftAssoc(d)` and define the method appropriately? The idea is that whenever we upgrade to a new version of Tasty, we inspect all the `hack*` methods and see if we can replace them with Tasty-Reflect calls.

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 3]
```diff
@@ -21,12 +21,20 @@ trait ClassLikeSupport:
...
     val (extensions, methods) = classDef.symbol.classMethods.filterNot(isSyntheticFunc).partition(isExtensionMethod(_))
 
-    val parsedExtensions = extensions
-      .map(parseMethod(_, isExtension = true))
+    
+    val parsedExtensions = extensions.map(e => (getExtendedSymbol(e).get, e) ).groupBy{
+      case (arg, e) => arg.symbol.pos
+    }.map{
+      case (pos, list) => if(list.size == 1) {
+        ExtensionGroup(parseArgument(list(0)._1, _ => "", true, false), List(parseMethod(list(0)._2, extInfo = Some(ExtensionInformation(false)))))
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: A note about `_1` and similar: in Scala 3 the preferred way to index a tuple is `(1)`, this is as efficient as `_1`.

**Review by abgruszecki — APPROVED** [review]
I have only minor nitpicks. Otherwise LGTM, but I couldn't really review the Dokka-interfacing parts well.

**Review by romanowski — APPROVED** [review]

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 3 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: It's a bit tricky as "_1" is actually "(0)" ;)

File: src/main/scala/dotty/dokka/tasty/SyntheticSupport.scala (hunk 2 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/SyntheticSupport.scala]: Done.

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala (hunk 1 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/ScalaPageCreator.scala]: Done

**Review by romanowski — APPROVED** [review]

**Pull request merged**